action.yml

```diff
@@ -75,59 +67,197 @@ runs:
+      id: modulemap
+      shell: bash
+      env:
+        GITHUB_REPOSITORY: ${{ github.repository }}
+      run: |
+        RESP_CODE=$(curl -w %{http_code} -s -L -o __repositories.json \
+          https://raw.githubusercontent.com/silverstripe/supported-modules/refs/heads/main/repositories.json
+        )
+        if [[ $RESP_CODE != "200" ]]; then
+          echo "Failed to fetch repositories.json - HTTP response code was $RESP_CODE"
+          cat __repositories.json
+          exit 1
+        fi
+        # Create a module map of { module_major: cms_version }
+        # e.g. for silverstripe/admin it will be { "1": "4", "2": "5", "3": "6" }
+        $MODULE_MAP=$(jq -r --arg github_repository "$GITHUB_REPOSITORY" '.supportedModules[] | select(.github == $github_repository) | .majorVersionMapping | to_entries | map({key: .value[0], value: .key}) | from_entries' __repositories.json)
+        echo "MODULE_MAP is $MODULE_MAP"
+        echo "module_map=$MODULE_MAP" >> $GITHUB_OUTPUT
+        rm __repositories.json
+
+    - name: Derive branches
+      id: branches
+      shell: bash
+      env:
+        GITHUB_REPOSITORY: ${{ github.repository }}
+        MAJOR_TYPE: ${{ steps.gettype.outputs.major_type }}
+        MINOR_TYPE: ${{ steps.gettype.outputs.minor_type }}
+        MODULE_MAP: ${{ steps.modulemap.outputs.module_map }}
+      run: |
+        MAJOR_BRANCH=""
+        MINOR_BRANCH=""
+        MAJOR_BRANCHES=$(jq -r '.[] | select(.name | test("^[0-9]+$")) | .name' __branches.json | sort -V -r)
         # "major_branches" needs to be quoted to preserve line-breaks
-        major_branches_count=$(echo "$major_branches" | wc -l)
-        if (( $major_branches_count < 1 )); then
+        MAJOR_BRANCHES_COUNT=$(echo "$MAJOR_BRANCHES" | wc -l)
+        if (( $MAJOR_BRANCHES_COUNT < 1 )); then
           echo "No major branches found, cannot continue"
           exit 1
         fi
-        latest_major_branch=$(echo "$major_branches" | head -1)
-        latest_major_stable_tags_count=0
+        LATEST_MAJOR_BRANCH=$(echo "$MAJOR_BRANCHES" | head -1)
+        LATEST_MAJOR_STABLE_TAGS_count=0
         # not escaping rx . as \. because jq will say it is a compile error
-        latest_major_stable_tags=$(jq -r ".[] | select(.name | test(\"^$latest_major_branch.[0-9]+.[0-9]+$\")) | .name" __tags.json | sort -V -r)
-        if [[ "$latest_major_stable_tags" != "" ]]; then
-          latest_major_stable_tags_count=$(echo "$latest_major_stable_tags" | wc -l)
-        fi
-        if [[ $major_type == "current" ]]; then
-          major_branch=$(echo "$major_branches" | head -1)
-          if (( $latest_major_stable_tags_count == 0 )) && (( $major_branches_count >= 2 )); then
-            major_branch=$(echo "$major_branches" | head -2 | tail -1)
+        LATEST_MAJOR_STABLE_TAGS=$(jq -r --arg latest_major_branch "$LATEST_MAJOR_BRANCH" ".[] | select(.name | test('^$latest_major_branch.[0-9]+.[0-9]+$')) | .name" __tags.json | sort -V -r)
```

Comment:
I've also refactored these lines to use the jq `--arg` flag which is a more robust way to pass in bash args